src/vs/workbench/browser/dnd.ts

```diff
@@ -590,6 +591,49 @@ export function fillEditorsDragData(accessor: ServicesAccessor, resourcesOrEdito
 
 //#endregion
 
+//#region DND contributions
+
+export interface IDragAndDropContributionRegistry {
+	/**
+	 * Registers a drag and drop contribution.
+	 */
+	register(contribution: IDragAndDropContribution): void;
+
+	/**
+	 * Returns all registered drag and drop contributions.
+	 */
+	getAll(): IterableIterator<IDragAndDropContribution>;
+}
+
+export interface IDragAndDropContribution {
+	readonly dataFormatKey: string;
+	getEditorInputs(data: string): IDraggedResourceEditorInput[];
+	setData(resources: IResourceStat[], event: DragMouseEvent | DragEvent): void;
+}
+
+class DragAndDropContributionRegistry implements IDragAndDropContributionRegistry {
+	private readonly _contributions = new Map<string, IDragAndDropContribution>();
+
+	register(contribution: IDragAndDropContribution): void {
+		if (this._contributions.has(contribution.dataFormatKey)) {
+			throw new Error(`A drag and drop contributiont with key '${contribution.dataFormatKey}' was already registered.`);
+		}
+		this._contributions.set(contribution.dataFormatKey, contribution);
+	}
+
+	getAll(): IterableIterator<IDragAndDropContribution> {
+		return this._contributions.values();
+	}
+}
+
+export const Extensions = {
+	DragAndDropContribution: 'base.contributions.dragAndDrop'
```

Comment:
Maybe more like `workbench.contributions.dragAndDrop` to align with the name of similar ones.